lab-abigail/server.js

```diff
@@ -0,0 +1,70 @@
+'use strict';
+
+const Client = require('./model/client.js');
+const net = require('net');
+const EE = require('events').EventEmitter;
+const ee = new EE();
+const server = module.exports = net.createServer();
+const PORT = process.env.PORT || 3000;
+
+server.listen(PORT, function() {
+  console.log(`Listening on: ${PORT}`);
+});
+
+const pool = [];
+
+ee.on('default', function(client, string) {
+  client.socket.write(`Not a valid command: ${string.split(' ', 1)}\n`);
+});
+
+ee.on('/all', function(client, string) {
+  pool.forEach(c => c.socket.write(`${client.nickName}: ${string}\n`));
+});
+
+ee.on('/nick', function(client, string) {
+  client.original = client.nickName;
+  client.nickName = string.trim().split(' ').join('').toUpperCase();
+  client.socket.write(`You changed your nickname to from ${client.original} to ${client.nickName}\n`);
+  pool.forEach(pool => {
+    pool.socket.write(`${client.original} changed his/her name to ${client.nickName}\n`);
+  });
+});
+
+ee.on('/dm', function(client, string) {
+  let nickName = string.split(' ').shift().trim().toUpperCase();
+  let message = string.split(' ').slice(1).join(' ').trim();
+
+  pool.forEach(dm => {
+    if (dm.nickName === nickName) {
+      dm.socket.write(`Private Message from ${client.nickName}: ${message}\n`);
+    }
+  });
+  client.socket.write(`You private messaged ${nickName}: ${message}\n`);
+});
+
+server.on('connection', socket => {
+  let client = new Client(socket);
+  pool.push(client);
+  pool.forEach(c => c.socket.write(`${client.nickName} has connected!\n`));
+
+  socket.on('data', data => {
+    let command = data.toString().split(' ').shift().trim();
+    // console.log(`${client.nickName}: ${data}`);
+    if(command.startsWith('/')) {
+      ee.emit(command, client, data.toString().split(' ').slice(1).join(' '));
+      return;
+    }
+    ee.emit('default', client, data.toString());
+  });
+
+
+  socket.on('error', error => {
+    console.log(`Error: ${error}`);
+  });
+
+  socket.on('close', data => {
```

Comment:
**TODO FOR LOST POINTS**
The lab directions also state this:
``When a socket emits the close event the socket should be removed from the client pool!``

This means that you should be iterating over your ``pool`` array and deleting that client from the pool. 

So without completely giving away the answer, you want to do something like...

```
pool.forEach(c => {
 //find a client in the pool that matches the current client (i.e. the client that shut down their connection). When found, delete them from this array
}
```

You  might want to look up docs on how the array method ``splice`` works to achieve this. See me if you're stuck.